lib/block_view_lockups.go

```diff
@@ -818,109 +1174,541 @@ func (bav *UtxoView) _connectCoinLockup(
+						existingLockedBalanceEntry.VestingEndTimestampNanoSecs)
+					if err != nil {
+						return 0, 0, nil,
+							errors.Wrap(err, "_connectCoinLockup failed to compute vested split")
+					}
+
+					// Consolidate the split and existing locked balance entry.
+					combinedBalanceBaseUnits, err := SafeUint256().Add(
+						&splitLockedBalanceEntry.BalanceBaseUnits,
+						&existingLockedBalanceEntry.BalanceBaseUnits)
+					if err != nil {
+						return 0, 0, nil,
+							errors.Wrap(RuleErrorCoinLockupYieldCausesOverflowInLockedBalanceEntry,
+								"_connectCoinLockup")
+					}
+					splitLockedBalanceEntry.BalanceBaseUnits = *combinedBalanceBaseUnits
+
+					// Set the now combined splitLockedBalanceEntry into the view.
+					bav._setLockedBalanceEntry(splitLockedBalanceEntry)
+
+					// Update the proposed locked balance entry with the remaining portion.
+					proposedLockedBalanceEntry = remainingLockedBalanceEntry
+
+					// (3b-iii) On the final iteration, the remaining proposedLockedBalanceEntry
+					//          is the only vesting schedule left.
+					if ii == len(lockedBalanceEntries)-1 {
+						bav._setLockedBalanceEntry(proposedLockedBalanceEntry)
+					}
+				}
+
+				// (3b-iv) By now, we know the edges to be trimmed as best possible.
+				// We check if the existing and proposed now overlap perfectly in time and combine if so.
+				if (existingLockedBalanceEntry.UnlockTimestampNanoSecs ==
+					proposedLockedBalanceEntry.UnlockTimestampNanoSecs) &&
+					(existingLockedBalanceEntry.VestingEndTimestampNanoSecs ==
+						proposedLockedBalanceEntry.VestingEndTimestampNanoSecs) {
+					// Combine the remaining balance.
+					combinedBalanceBaseUnits, err := SafeUint256().Add(
+						&existingLockedBalanceEntry.BalanceBaseUnits,
+						&proposedLockedBalanceEntry.BalanceBaseUnits)
+					if err != nil {
+						return 0, 0, nil,
+							errors.Wrap(RuleErrorCoinLockupYieldCausesOverflowInLockedBalanceEntry,
+								"_connectCoinLockup")
+					}
+
+					// Update the remaining entry.
+					proposedLockedBalanceEntry.BalanceBaseUnits = *combinedBalanceBaseUnits
+					bav._setLockedBalanceEntry(proposedLockedBalanceEntry)
+				}
+			}
+		}
+	}
 
 	// Add a UtxoOperation for easy reversion during disconnect.
 	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
 		Type:                       OperationTypeCoinLockup,
 		PrevTransactorBalanceEntry: prevTransactorBalanceEntry,
-		PrevLockedBalanceEntry:     &previousLockedBalanceEntry,
+		PrevLockedBalanceEntry:     previousLockedBalanceEntry,
 		PrevCoinEntry:              prevCoinEntry,
 	})
 
 	// Construct UtxoOps in the event this transaction is reverted.
 	return totalInput, totalOutput, utxoOpsForTxn, nil
 }
 
+func SplitVestedLockedBalanceEntry(
+	lockedBalanceEntry *LockedBalanceEntry,
+	startSplitTimestampNanoSecs int64,
+	endSplitTimestampNanoSecs int64,
+) (
+	_splitLockedBalanceEntry *LockedBalanceEntry,
+	_remainingLockedBalanceEntry *LockedBalanceEntry,
+	_err error,
+) {
```

Comment:
Nit: One note I had while reading through this: It seems like a better API for this would be to pass it a single `splitTimestamp` and then it gives you back `leftBalanceEntry` and `rightBalanceEntry`. I also felt like that would make it a bit easier to read above, since "left" is something you can visualize, whereas "remaining" isn't. I don't think you need to redo it, but maybe leave a comment if you leave it as-is.